benchmark/C/Savina/src/parallelism/GuidedSearch.lf

```diff
@@ -0,0 +1,246 @@
+        "GuidedSearch/GridNode.hh"],
+    logging: "warn"
+};
+
+preamble {=
+    #include "GuidedSearchCommon.hh"
+    #include <deque>
+    #include <list>
+    #include <iostream>
+=}
+
+reactor Manager(numWorkers: size_t(20), gridSize: size_t(30), priorities: size_t(30)) {
+
+    state grid: Grid*;
+    state nodesToVisit: std::deque<GridNode*>;
+    state pathWasFound: bool(false);
+    
+    input start: bool;
+    output finished: bool;
+    
+    output[numWorkers] search: WorkMessage;
+    input[numWorkers] pathFound: bool;
+    input[numWorkers] moreNodesToVisit: std::deque<GridNode*>;
+    
+    logical action next;
+    
+    reaction (start) -> next {=
+        self->grid = new Grid(self->gridSize, self->priorities);
+        self->nodesToVisit = std::deque<GridNode*>();
+        srand(123456);
+        self->grid->initializeData(); // reset the self->grid
+        self->nodesToVisit.clear();
+        self->pathWasFound = false;
+        
+        // start execution, first visit the origin node
+        self->nodesToVisit.push_back(self->grid->originNode());
+        schedule(next, 0);
+    =}
+    
+    reaction (next) -> next, search {=
+        // abort if path was found
+        if (self->pathWasFound) {
+            return;
+        }
+
+        // send nodes from the internal queue to all workers
+        size_t workerIndex{0};        
+        while(!self->nodesToVisit.empty() && workerIndex < self->numWorkers) {
+            GridNode* startNode = self->nodesToVisit.front();
+            self->nodesToVisit.pop_front();
+            SET(search[workerIndex], WorkMessage(startNode, self->grid->targetNode()));
+            workerIndex++;
+        }
+        
+         // continue with the next iteration
+         schedule(next, 0);
+    =}
+    
+    reaction(pathFound) -> finished {=
+        self->pathWasFound = true;
+        bool valid = self->grid->validate();
+        info_print("Found a path through the grid");
+        info_print("Result valid: %d", valid);
+        SET(finished, true);
+    =}
+    
+    reaction (moreNodesToVisit) {=
+        // collect more nodes to visit from all workers and store them in the internal queue
+        for (int i=0; i< moreNodesToVisit_width; i++) {
+            if (moreNodesToVisit[i]->is_present) {
+                auto nodes = moreNodesToVisit[i]->value;
+                if (!nodes.empty()) {
+                    self->nodesToVisit.insert(self->nodesToVisit.end(), nodes.begin(), nodes.end());
+                }
+            }
+        }
+    =}
+}
+
+reactor Worker(bank_index: int(0), threshold: size_t(1024)) {
+    
+    input search: WorkMessage;
+    output pathFound: bool;
+    output moreNodesToVisit: std::deque<GridNode*>;
+
+    preamble {=
+
+        void busyWait() {
+            for (int i = 0; i < 100; i++) {
+                rand();
+            }
+        }
+    =}
+    
+    reaction(search) -> pathFound, moreNodesToVisit {=
+        auto workMessage = search->value;
```

Comment:
Also this copies (although it's only two pointers and will likely not matter)

Reply:
I think it did have a slight performance benefit to use smart pointers.